MTM-46010 Added new section for microfrontends. (#561)
SoftwareAG/c8y-docs: content/web/microfrontends-bundle/add-a-new-package-as-blueprint.md

---
title: Custom package blueprint with Module Federation
layout: redirect
weight: 40
---

**Version:** 1011.153.0 | **Packages:** @c8y/cli, @c8y/apps and @c8y/ngx-components

> Blueprints are composition of multiple UI functionalities that can be hosted by the platform (static files) and can be used to scaffold a new solution from scratch. On the other hand, a package is the composition of plugins and blueprints. As a blueprint can export plugins as well, they can be packed together into one package and deployed to the platform. 

Reviewer: composition > in the user guide I changed it to "combination" for clarity; Not sure if this is a technical term we should leave here in this context?

Author: combination is fine

> ### 1. Initialize the package blueprint example

Reviewer: package blueprint > unclear use of these terms

> Use the command shown below to start a multi-step process of creating a sample blueprint:

Reviewer: ```suggestion
Use the command shown below to start a multi-step process for creating a sample blueprint:
```

```js
c8ycli new
```

> Select the plugin name, for example, "package-blueprint":

Reviewer: Usually we use numbered lists for such step by step procedures. In the other tutorials we even add them as headings. Please compare and make it consistent.

Author: added numberings

```

Select the version for which you want to create a sample application, for example, "1015.33.0 (next)":

```console
? Which base version do you want to scaffold from? (Use arrow keys)
  1011.0.30 (latest)
> 1015.33.0 (next)
  1015.0.39
  1014.0.104
  1013.0.193
  1010.0.38
  other
```

Select an application template as the basis for your plugin, for example, "package-blueprint":

```console
? Which base project do you want to scaffold from?
  cockpit
  devicemanagement
  hybrid
> package-blueprint
  tutorial
  widget-plugin
  administration
```

After a few seconds, you should see the following message:

```console
Application created. Go into the folder "app-blueprint" and run npm install
```

Navigate to your application folder and execute `npm install`.

The application folder should look like the example shown below.
For this tutorial, the most important files are *package.json* and `README.md`.

```console
app.module.spec.ts;
jest.config.js;
README.md;
tsconfig.spec.json;
app.module.ts;
package.json;
setup-jest.js;
index.ts;
polyfills.ts;
tsconfig.json;
```

You have now created your first package blueprint that uses Module Federation.

> ### 2. Differences in approach to creating custom applications

Reviewer: remove the number from the heading, we use numbers for procedural steps only

Author: removed

There are a couple of differences between a simple widget and one that is built according to the Module Federation guidelines.

> The biggest difference is the *package.json* file, where fields such as `isPackage` and `package` and `exports` (not in the current blueprint app) are located.

Reviewer: ```suggestion
The biggest difference is the *package.json* file, where fields such as `isPackage` and `package` and `exports` (not in the current blueprint app) are located.
```
```suggestion
The largest difference can be found in the *package.json* file, where fields such as `isPackage` and `package` and `exports` (not in the current blueprint app) are located.
```

The following list shows the fields and what they are responsible for:

- `isPackage`: Indicates if the application is a package. In case of a widget that is added using Module Federation, set the value to `true`.

> - `package`: The type of package (for example, `blueprint`, but the type of the package can also be a plugin.).

Reviewer: ```suggestion
- `package`: The type of package (for example, `blueprint`, but the type of the package can also be a plugin).
```

> - `exports`: Important field. Defines the Angular modules that will be made available by the widget-plugin for the shell application (see also the *README.md* file):

Reviewer: remove colon?

Author: removed

  - `name`: The name of the exported module.
  - `module`: The name of the Angular module class.
  - `path`: The path to the TypeScript file with the module.
  - `description`: A brief description of what the module does.

{{< c8y-admon-info >}}
A blueprint can also include plugins, which can later be used to extend other applications.
{{< /c8y-admon-info >}}

> There is also a difference in how to start the local development server, see the following step for more information on the server's role.

Reviewer 2: Where is this part of the local development server?

Author: removed this part as it was not relevant for this case.

---
title: Custom package blueprint with Module Federation
layout: redirect
weight: 40
---

**Version:** 1015.33.0 | **Packages:** @c8y/cli, @c8y/apps and @c8y/ngx-components

With Module Federation it is possible to add new functionality while the application is **running** (runtime), whereas the old approach only allowed new functionality to be added before the application was **built** (compile time).

> Blueprints are combination of multiple UI functionalities that can be hosted by the platform (static files) and can be used to scaffold a new solution from scratch. On the other hand, a package is the composition of plugins and blueprints. As a blueprint can export plugins as well, they can be packed together into one package and deployed to the platform.

Reviewer: ```suggestion
Blueprints are combinations of multiple UI functionalities that can be hosted by the platform (static files) and can be used to scaffold a new solution from scratch. On the other hand, a package is the composition of plugins and blueprints. As a blueprint can export plugins as well, they can be packed together into one package and deployed to the platform.
```

### Initialize the blueprint example

1. Use the command shown below to start a multi-step process of creating a sample blueprint:

```js
c8ycli new
```

2. Select the plugin name, for example, "package-blueprint":

```js
? Enter the name of the project:  (my-application) package-blueprint
```

3. Select the version for which you want to create a sample application, for example, "1015.33.0 (next)":

```console
? Which base version do you want to scaffold from? (Use arrow keys)
  1011.0.30 (latest)
> 1015.33.0 (next)
  1015.0.39
  1014.0.104
  1013.0.193
  1010.0.38
  other
```

4. Select an application template as the basis for your plugin, for example, "package-blueprint":

```console
? Which base project do you want to scaffold from?
  cockpit
  devicemanagement
  hybrid
> package-blueprint
  tutorial
  widget-plugin
  administration
```

After a few seconds, you should see the following message:

```console
Application created. Go into the folder "app-blueprint" and run npm install
```

5. Navigate to your application folder and execute `npm install`.

The application folder should look like the example shown below.

> For this tutorial, the most important files are _package.json_ and `README.md`.

Reviewer: ```suggestion
For this tutorial, the most important files are *package.json* and *README.md*.
```

```console
app.module.spec.ts;
jest.config.js;
README.md;
tsconfig.spec.json;
app.module.ts;
package.json;
setup-jest.js;
index.ts;
polyfills.ts;
tsconfig.json;
```

You have now created your first package blueprint that uses Module Federation.

### Stepper setup (optional)

> The HOOK_STEPPER can be additionally provided to allow application customization during the first load of an applicaiton. In this optional step we will show a small single step example in which the user will be able to choose whether the navigator will be collapsed or not on start up.

Reviewer: ```suggestion
The HOOK_STEPPER can be additionally provided to allow application customization during the first load of an application. In this optional step we show a small single step example in which the user can select whether the navigator will be collapsed or not on start up.
```

> 1. Create a new "setup-step1.component.ts" file with the following content:

Reviewer: ```suggestion
1. Create a new *setup-step1.component.ts* file with the following content:
```

---
title: Custom package blueprint with Module Federation
layout: redirect
weight: 40
---

> **Version:** 1011.153.0 | **Packages:** @c8y/cli, @c8y/apps and @c8y/ngx-components

Reviewer 2: Can you update the version? I guess it is not that old, or is it?

> Blueprints are combination of multiple UI functionalities that can be hosted by the platform (static files) and can be used to scaffold a new solution from scratch. On the other hand, a package is the composition of plugins and blueprints. As a blueprint can export plugins as well, they can be packed together into one package and deployed to the platform.

Reviewer 2: maybe it is good to mention what module federation is, as you mention it in the title?

    protected step: CdkStep,
    protected setup: SetupComponent,
    protected appState: AppStateService,
    protected alert: AlertService
  ) {
    super(stepper, step, setup, appState, alert);
  }
}
```

2. Create a "setup-step1.component.html" template:

```html
<form #stepForm="ngForm" name="form" class="d-contents">
  <div class="container-fluid flex-no-shrink fit-w">
    <div class="row separator-bottom">
      <div
        class="col-md-8 col-md-offset-2 col-lg-6 col-lg-offset-3 p-t-24 p-l-16 p-r-16"
      >
        <h3 translate class="text-medium l-h-base">Misc</h3>
        <p class="lead text-normal" translate>
          Miscellaneous settings for the current application.
        </p>
      </div>
    </div>
  </div>
  <div class="inner-scroll flex-grow">
    <div class="container-fluid fit-w">
      <div class="row">
        <div class="col-md-8 col-md-offset-2 col-lg-6 col-lg-offset-3">
          <c8y-misc-config [config]="config"></c8y-misc-config>
        </div>
      </div>
    </div>
  </div>
  <c8y-cockpit-setup-stepper-buttons
    [index]="stepper.selectedIndex"
    (onNext)="next()"
    (onBack)="back()"
  >
  </c8y-cockpit-setup-stepper-buttons>
</form>
```

> 3. Add a "cockpit-setup-stepper-buttons.component.ts" file which would show the buttons on each step:

Reviewer 2: I guess this can be omitted and the template on line 137 to 142 can include directly this buttons.

```typescript
import {Component, EventEmitter, Input, Output} from "@angular/core";

@Component({
  selector: "c8y-cockpit-setup-stepper-buttons",
  templateUrl: "./cockpit-setup-stepper-buttons.component.html"
})
export class CockpitSetupStepperButtonsComponent {
  @Input() index;
  @Output() onNext = new EventEmitter<void>();
  @Output() onBack = new EventEmitter<void>();
}
```

4. Add a "cockpit-setup-stepper-buttons.component.html" template:

```html
<div class="card-footer separator d-flex j-c-center">
  <button
    class="btn btn-default"
    type="button"
    (click)="onBack.emit()"
    *ngIf="index !== 0"
    translate
  >
    Previous
  </button>
  <button
    class="btn btn-primary"
    type="submit"
    (click)="onNext.emit()"
    translate
  >
    Save and continue
  </button>
</div>
```

5. Add a "setup-step.ts" file in which we will define the default app configuration:

```typescript
import {CdkStep} from "@angular/cdk/stepper";
import {
  AppStateService,
  AlertService,
  C8yStepper,
  SetupComponent
} from "@c8y/ngx-components";

> export abstract class SetupStep {

Reviewer 2: you could also directly add this to the component you defined in 1. This was only done to keep it more DRY (don't repeat yourself) but for this example, it makes it just too complex.
